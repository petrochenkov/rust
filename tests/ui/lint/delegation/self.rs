#![deny(delegations_detailed, delegations_per_parent_stats)]

struct F;

impl F {
    fn bar<T>(x: T) -> T { x }
    pub fn foo<T>(self, x: T) -> T { x }
    fn get_self() -> F { F }
}

struct S {
    field: F,
}

struct Stuple(F);

impl S {
    fn bar<T>(x: T) -> T {
        F::bar::<T>(x)
        //~^ ERROR dstats. parent: InherentImpl, stmts: ZeroWithTail, args_match: Same, args_preproc: false, ret_match: Same, callee_has_self: false, caller_has_self: false, same_name: true, ret_postproc: false.
    }

    fn foo<T>(self, x: T) -> T {
        self.field.foo::<T>(x)
        //~^ ERROR dstats. parent: InherentImpl, stmts: ZeroWithTail, args_match: Same, args_preproc: false, ret_match: Same, callee_has_self: true, caller_has_self: true, same_name: true, ret_postproc: false.
    }
}

fn foo<T>(x: T) -> T {
    F::bar::<T>(x)
    //~^ ERROR dstats. parent: Other, stmts: ZeroWithTail, args_match: Same, args_preproc: false, ret_match: Same, callee_has_self: false, caller_has_self: false, same_name: false, ret_postproc: false.
}

fn bar(x: i32) {
    F::get_self().foo::<i32>(x);
    //~^ ERROR dstats. parent: Other, stmts: OneWithoutTail, args_match: Different, args_preproc: true, ret_match: Different, callee_has_self: true, caller_has_self: false, same_name: false, ret_postproc: true.
    //~| ERROR dstats. parent: Other, stmts: OneWithoutTail, args_match: Different, args_preproc: false, ret_match: Different, callee_has_self: false, caller_has_self: false, same_name: false, ret_postproc: true.
}

trait Trait {
    fn foo<T>(x: T) -> T {
        foo(x)
        //~^ ERROR dstats. parent: Trait, stmts: ZeroWithTail, args_match: Same, args_preproc: false, ret_match: Same, callee_has_self: false, caller_has_self: false, same_name: true, ret_postproc: false.
    }
}

fn main() {}
