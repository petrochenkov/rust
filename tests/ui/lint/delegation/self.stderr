error: caller_parent: Other, stmts_before: false, arg0_match: Same, arg0_preproc: No, args_match: Same, args_preproc: No, ret_match: Same, has_self: false, caller_has_self: false, same_name: false, ret_postproc: false, parent: InherentImpl, source: User
  --> $DIR/self.rs:30:5
   |
LL | / fn foo<T>(x: T) -> T {
LL | |     F::bar::<T>(x)
   | |     ^^^^^^^^^^^^^^ callee
LL | |
LL | | }
   | |_- caller
   |
note: the lint level is defined here
  --> $DIR/self.rs:1:9
   |
LL | #![deny(delegations_detailed, delegations_per_parent_stats)]
   |         ^^^^^^^^^^^^^^^^^^^^

error: caller_parent: Other, stmts_before: false, arg0_match: Different, arg0_preproc: Other, args_match: DifferentCount, args_preproc: Other, ret_match: Different, has_self: true, caller_has_self: false, same_name: false, ret_postproc: true, parent: InherentImpl, source: User
  --> $DIR/self.rs:35:5
   |
LL | / fn bar(x: i32) {
LL | |     F::get_self().foo::<i32>(x);
   | |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ callee
LL | |
LL | |
LL | | }
   | |_- caller

error: caller_parent: Other, stmts_before: false, arg0_match: Different, arg0_preproc: Other, args_match: DifferentCount, args_preproc: Other, ret_match: Different, has_self: false, caller_has_self: false, same_name: false, ret_postproc: true, parent: InherentImpl, source: User
  --> $DIR/self.rs:35:5
   |
LL | / fn bar(x: i32) {
LL | |     F::get_self().foo::<i32>(x);
   | |     ^^^^^^^^^^^^^ callee
LL | |
LL | |
LL | | }
   | |_- caller

error: caller_parent: Trait, stmts_before: false, arg0_match: Same, arg0_preproc: No, args_match: Same, args_preproc: No, ret_match: Same, has_self: false, caller_has_self: false, same_name: true, ret_postproc: false, parent: Other, source: User
  --> $DIR/self.rs:42:9
   |
LL | /     fn foo<T>(x: T) -> T {
LL | |         foo(x)
   | |         ^^^^^^ callee
LL | |
LL | |     }
   | |_____- caller

error: caller_parent: InherentImpl, stmts_before: false, arg0_match: Same, arg0_preproc: No, args_match: Same, args_preproc: No, ret_match: Same, has_self: false, caller_has_self: false, same_name: true, ret_postproc: false, parent: InherentImpl, source: User
  --> $DIR/self.rs:19:9
   |
LL | /     fn bar<T>(x: T) -> T {
LL | |         F::bar::<T>(x)
   | |         ^^^^^^^^^^^^^^ callee
LL | |
LL | |     }
   | |_____- caller

error: caller_parent: InherentImpl, stmts_before: false, arg0_match: Different, arg0_preproc: Field, args_match: Same, args_preproc: No, ret_match: Same, has_self: true, caller_has_self: true, same_name: true, ret_postproc: false, parent: InherentImpl, source: User
  --> $DIR/self.rs:24:9
   |
LL | /     fn foo<T>(self, x: T) -> T {
LL | |         self.field.foo::<T>(x)
   | |         ^^^^^^^^^^^^^^^^^^^^^^ callee
LL | |
LL | |     }
   | |_____- caller

error: delegation count: 1, parent count: 2
   |
note: the lint level is defined here
  --> $DIR/self.rs:1:31
   |
LL | #![deny(delegations_detailed, delegations_per_parent_stats)]
   |                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: delegation count: 2, parent count: 1

error: aborting due to 8 previous errors

